### DISCUSSION TITLE
Add frozen module to unix port

### DISCUSSION THREAD

**Discussion by Krischan101** [micropython/micropython#13286] [Other] [answered]
Hi everyone,
I checked the the documentation, but I couldn't find away to add a custom frozen modules to unix port. Also, I don't see a modules folder in unix port.

**Comment by andrewleech** [comment]
The standard way to freeze modules (on all ports) is a manifest file: https://docs.micropython.org/en/latest/reference/manifest.html

The modules folder is the old approach that only worked on a few ports.

**Reply by Krischan101** [reply]
where should I place this manifest file?

I'm still new to micropython.

**Comment by shariltumin** [comment, marked as answer]
This is the way that I have done it for my unix port:

1. cd micropython
2. make a directory under micropython for example "extra". mkdir extra
3. put all your python modules under "micropython/extra"
4. cd ports/unix
5. edit "variants/manifest.py"
6. make V=1
7. cd build-standard/
8. ./micropython

Here is the content of my "variants/manifest.py" file

```
require("mip-cmdline")
# my extra modules
freeze('$(MPY_DIR)/extra', 'CryptoXo.py')
freeze('$(MPY_DIR)/extra', 'Caesar.py')
freeze('$(MPY_DIR)/extra', 'dotdict.py')
# -freeze('$(MPY_DIR)/extra', 'PickleDB.py')
freeze('$(MPY_DIR)/extra', 'pipe.py')
freeze('$(MPY_DIR)/extra', 'tscp.py')
freeze('$(MPY_DIR)/extra', 'worker.py')
# -freeze('$(MPY_DIR)/extra', 'wrepl.py')
```

Now ```worker``` is part of micropython, but ```wrepl``` is not.

```
$ ./micropython 
MicroPython v1.20.0-555-g92717a95c-dirty on 2023-12-28; linux [GCC 9.4.0] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import worker
>>> import wrepl
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: no module named 'wrepl'
>>>
```

**Reply by Krischan101** [reply]
Wow, thank you for this.

**Reply by jimmo** [reply]
Similar to the reply I just posted here: https://github.com/orgs/micropython/discussions/13337#discussioncomment-8042679

Rather than modifying anything in the MicroPython repo, customisations to the build should be done by creating your own board definition. On the unix port, these are called "variants".

You can build with your custom variant with

```
make VARIANT_DIR=../../path/to/variant
```

I suggest copying e.g. the `standard` variant as a starting point. You might need to change the include path from `../mpconfigvariant_common.h` to `boards/mpconfigvariant_common.h` if you want to re-use the common config.
